File: microservice-architecture/api-versioning/ts/test/test_router.ts
import { productHandlerV1, productHandlerV2, Request, VersionedRouter } from "../src/router";

function assertEqual<T>(expected: T, actual: T, message: string): void {
  if (expected !== actual) {
    throw new Error(`${message} expected=${expected} actual=${actual}`);
  }
}

const router = new VersionedRouter("v1");
router.register("v1", productHandlerV1);
router.register("v2", productHandlerV2);

const defaultRequest: Request = { path: "/products/P100", headers: {} };
assertEqual("v1", router.handle(defaultRequest).version, "default version should be v1");

const headerRequest: Request = { path: "/products/P100", headers: { "X-API-Version": "2" } };
assertEqual("v2", router.handle(headerRequest).version, "header should select v2");

const unsupportedRequest: Request = { path: "/products/P100", headers: { "X-API-Version": "9" } };
assertEqual(400, router.handle(unsupportedRequest).statusCode, "unsupported version should fail");

console.log("api-versioning(ts) tests passed");

File: microservice-architecture/api-versioning/python/test/test_router.py
import os
import sys
import unittest

sys.path.append(os.getcwd())

from src.router import (  # noqa: E402
    Request,
    VersionedRouter,
    product_handler_v1,
    product_handler_v2,
)


class ApiVersioningTest(unittest.TestCase):
    def setUp(self) -> None:
        self.router = VersionedRouter("v1")
        self.router.register("v1", product_handler_v1)
        self.router.register("v2", product_handler_v2)

    def test_default_version_falls_back_to_v1(self) -> None:
        response = self.router.handle(Request("/products/P100", {}))
        self.assertEqual("v1", response.version)
        self.assertIn('"name":"Mechanical Keyboard"', response.body)

    def test_header_can_select_v2(self) -> None:
        response = self.router.handle(Request("/products/P100", {"X-API-Version": "2"}))
        self.assertEqual("v2", response.version)
        self.assertIn('"inventoryStatus":"IN_STOCK"', response.body)

    def test_unsupported_version_returns_bad_request(self) -> None:
        response = self.router.handle(Request("/products/P100", {"X-API-Version": "9"}))
        self.assertEqual(400, response.status_code)


if __name__ == "__main__":
    unittest.main()
